### PR TITLE
Add Async API Tests for Timeout

### PR DESCRIPTION
**Pull request by cheenamalhotra**
All tests passing.

### CODE REVIEW

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/AsyncTest/AsyncTimeoutTest.cs [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+        static string delayQuery10s = "WAITFOR DELAY '00:00:10'";
+
+        public enum AsyncAPI
+        {
+            ExecuteReaderAsync,
+            ExecuteScalarAsync,
+            ExecuteXmlReaderAsync
+        }
+
+        [ConditionalTheory(typeof(DataTestUtility), nameof(DataTestUtility.AreConnStringSetupForAE)]
```
Comment (cheenamalhotra) [inline src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/AsyncTest/AsyncTimeoutTest.cs]:
```suggestion
        [ConditionalTheory(typeof(DataTestUtility), nameof(DataTestUtility.AreConnStringsSetup))]
```

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/AsyncTest/AsyncTimeoutTest.cs [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+            {
+                sqlConnection.Open();
+                if (timeoutDelay != 0)
+                {
+                    ConnectionHelper.SetEnforcedTimeout(sqlConnection, true, timeoutDelay);
+                }
+                switch (commonObj)
+                {
+                    case "Connection":
+                        QueryAndValidate(api, 1, delayQuery2s, 1, true, true, sqlConnection).Wait(); // Waith's PR
```
Comment (cheenamalhotra) [inline src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/AsyncTest/AsyncTimeoutTest.cs]:
```suggestion
                        QueryAndValidate(api, 1, delayQuery2s, 1, true, true, sqlConnection).Wait();
```

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/AsyncTest/AsyncTimeoutTest.cs [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+                    case "Connection":
+                        QueryAndValidate(api, 1, delayQuery2s, 1, true, true, sqlConnection).Wait();
+                        QueryAndValidate(api, 2, delayQuery2s, 5, false, true, sqlConnection).Wait();
+                        QueryAndValidate(api, 3, delayQuery10s, 1, true, true, sqlConnection).Wait();
+                        QueryAndValidate(api, 4, delayQuery2s, 10, false, true, sqlConnection).Wait();
+                        break;
+                    case "Command":
+                        using (SqlCommand cmd = sqlConnection.CreateCommand())
+                        {
+                            QueryAndValidate(api, 1, delayQuery2s, 1, true, false, sqlConnection, cmd).Wait(); // Waith's PR
```
Comment (cheenamalhotra) [inline src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/AsyncTest/AsyncTimeoutTest.cs]:
```suggestion
                            QueryAndValidate(api, 1, delayQuery2s, 1, true, false, sqlConnection, cmd).Wait();
```

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParserStateObject.cs [hunk 4]
```diff
@@ -125,8 +125,14 @@ internal enum SnapshottedStateFlags : byte
         internal volatile bool _attentionSent;              // true if we sent an Attention to the server
         internal volatile bool _attentionSending;
 
-        private readonly LastIOTimer _lastSuccessfulIOTimer;                     
-                                                                                 
+        // Below 2 properties are used to enforce timeout delays in code to 
+        // reproduce issues related to theadpool starvation and timeout delay.
+        // It should always be set to false by default, and only be enabled during testing.
+        internal bool _enforceTimeoutDelay = true;
```
Comment (cheenamalhotra) [inline src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParserStateObject.cs]:
```suggestion
        internal bool _enforceTimeoutDelay = false;
```

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParserStateObject.cs [hunk 5]
```diff
@@ -3560,35 +3568,35 @@ internal void SendAttention(bool mustTakeWriteLock = false)
...
                         {
-                            if (releaseLock)
-                            {
-                                _parser.Connection.ThreadHasParserLockForClose = false;
-                                _parser.Connection._parserLock.Release();
-                            }
+                            _parser.Connection.ThreadHasParserLockForClose = false;
+                            _parser.Connection._parserLock.Release();
                         }
+                    }
```
Comment (Wraith2) [inline src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParserStateObject.cs]: Would you mind reverting this indentation change? The IDE isn't seeing inside the #ifdef's and is taking the block 1 step too far left as a result.

Comment (cheenamalhotra) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParserStateObject.cs]: I've fixed the brackets, so they won't bother again.

**Review by Wraith2 — APPROVED** [review]

**Pull request merged**